client/beefy/src/worker.rs

```diff
@@ -216,77 +241,225 @@ where
+
+				// id is stored for skipped session metric calculation
+				self.last_signed_id = rounds.validator_set_id();
+
+				let block_num = round.1;
+				let commitment = Commitment {
+					payload: round.0,
+					block_number: block_num,
+					validator_set_id: self.last_signed_id,
+				};
+
+				let signed_commitment = SignedCommitment { commitment, signatures };
+
+				metric_set!(self, beefy_round_concluded, block_num);
+
+				info!(target: "beefy", "🥩 Round #{} concluded, committed: {:?}.", round.1, signed_commitment);
+
+				if self
+					.backend
+					.append_justification(
+						BlockId::Number(block_num),
+						(
+							BEEFY_ENGINE_ID,
+							VersionedFinalityProof::V1(signed_commitment.clone()).encode(),
+						),
+					)
+					.is_err()
+				{
+					// just a trace, because until the round lifecycle is improved, we will
+					// conclude certain rounds multiple times.
+					trace!(target: "beefy", "🥩 Failed to append justification: {:?}", signed_commitment);
+				}
+				self.signed_commitment_sender
+					.notify(|| Ok::<_, ()>(signed_commitment))
+					.expect("forwards closure result; the closure always returns Ok; qed.");
+
+				self.set_best_beefy_block(block_num, None);
+
+				trace!(target: "beefy", "🥩 round concluded, waking up ticker");
+				// We can ignore 'buffer full' error since it means some other task already
+				// queued up a wake message.
+				let _ = self.waker_tx.try_send(());
+			}
+		}
+	}
+
+	/// Set best BEEFY block to `block_num`.
+	///
+	/// Also sends/updates the best BEEFY block hash to the RPC worker.
+	/// If the hash is not explicitly provided in `opt_block_hash`,
+	/// this function just gets it from the `Client`.
+	fn set_best_beefy_block(
+		&mut self,
+		block_num: NumberFor<B>,
+		mut opt_block_hash: Option<<<B as Block>::Header as Header>::Hash>,
+	) {
+		if opt_block_hash.is_none() {
+			// Try to get block hash ourselves.
+			opt_block_hash = match self.client.hash(block_num) {
+				Ok(h) => h,
+				Err(e) => {
+					error!(target: "beefy", "🥩 Failed to get hash for block number {}: {}",
+						block_num, e);
+					None
+				},
+			}
+		};
+		// Update RPC worker with new best BEEFY block hash.
+		opt_block_hash.map(|hash| {
+			self.beefy_best_block_sender
+				.notify(|| Ok::<_, ()>(hash))
+				.expect("forwards closure result; the closure always returns Ok; qed.")
+		});
+		// Set new best BEEFY block number.
+		self.best_beefy_block = Some(block_num);
+		metric_set!(self, beefy_best_block, block_num);
+	}
+
+	/// Return first block number pertaining to same session as `header`.
+	fn session_start_for_header(&self, header: &B::Header) -> NumberFor<B> {
+		// Current header is also the session start.
+		if let Some(_) = find_authorities_change::<B>(header) {
+			return *header.number()
+		}
+
+		// Walk up the chain looking for the session change digest.
+		let mut parent_hash = *header.parent_hash();
+		while parent_hash != Default::default() {
```

Comment:
Just periodically polling and making Runtime calls which are expensive, hence I don't think this is the best solution and it's better to resolve to some sort of heuristic.

Reply:
Since avg session length is runtime-specific, it would have to be exposed through some new runtime api. Then we're back to expensive Runtime calls.

What if we just check BEEFY pallet when starting the worker?
- `not present` -> slow periodic poll (1min / 5min) waiting for pallet instantiation (`Some(beefy-validator-set)` coming from `BeefyApi`)
- once BEEFY pallet is `present` -> we can start worker and safely walk the chain each time we witness a validators-set-change, knowing there's a guaranteed `AuthoritiesChange` digest associated with it.

Reply:
IMO this is working backwards, we shouldn't be in this situation in the first place. If the validator set has changed then it must have been because some previous block had the `AuthoritiesChange` digest, IMO we should have found it when getting the finality notification for that block and noted it.

Edit: I guess there would be some edge case for startup, didn't think about it.